strophe.h

```diff
@@ -295,6 +295,8 @@ xmpp_ctx_t *xmpp_conn_get_context(xmpp_conn_t *conn);
 void xmpp_conn_disable_tls(xmpp_conn_t *conn);
 int xmpp_conn_is_secured(xmpp_conn_t *conn);
 void xmpp_conn_set_keepalive(xmpp_conn_t *conn, int timeout, int interval);
+void xmpp_conn_set_keepalive_full(
+    xmpp_conn_t *conn, int timeout, int interval, int count, int user_timeout);
```

Comment:
I really like the idea of supporting those options, but I'm not sure if that API is a good idea ... I'm already seeing a `xmpp_conn_set_keepalive_fuller()` and `xmpp_conn_set_keepalive_but_this_time_really_full()` in a not so distant future ;)

I guess we should go with something better compatible with the `setsockopt()` capabilities and take into account the possibility that not only one socket may be used in the future.

Thinking a bit about it I see these two options, but I'm sure there are more and better ones?

1. add an API that allows setting values dynamically ... maybe something like this? (This proposal supports multiple values at once, a single-value API would also be possible)

```c
/* This abstraction allows us to map internally to `level` + `optname` */
typedef enum {
  /* those names are not final yet ... */
  XMPP_KA_TIMEOUT = 0, /* maybe this should be `XMPP_SO_TIMEOUT` or `XMPP_SO_SOC_TIMEOUT` */
  XMPP_KA_INTERVAL, /* ... and this one `XMPP_TCP_INTERVAL` or `XMPP_SO_TCP_INTERVAL` ... */
  ...
} xmpp_sockopt_en;

typedef struct {
  xmpp_sockopt_en type;
  /* this allows us to support the full set of options, but also have an easy-to-use API for int values */
  union {
    void *p;
    int i;
  };
  size_t p_len; /* if != 0, use `p`, else use `i` */
} xmpp_sockopt_t;

int xmpp_conn_set_socket_options(xmpp_conn_t *conn, xmpp_sockopt_t *arr, size_t arrsz);
```

2. define a call-back way and let the user do the work

```c
/** `sock` is of type `SOCKET` on Win and `int` on the others */
typedef int (*xmpp_sockopt_cb)(void *sock);
void xmpp_conn_set_sockopt_callback(xmpp_conn_t *conn, xmpp_sockopt_cb *cb);
```

Where 2 has 

* the clear advantage that we don't have to take care if something changes under the hood, new API's or options to e.g. `setsockopt()` are added.
* the clear disadvantage that the user now has to do everything themselves and they maybe don't have the knowledge.

I'm slightly in favor of something like 1 as we're trying to be a user-friendly library and users shouldn't have to have that deep networking knowledge.

Regarding Windows support: I guess we're fine here for now with the existing API and it looks better for the future https://docs.microsoft.com/en-us/windows/win32/winsock/ipproto-tcp-socket-options

Reply:
I agree `xmpp_conn_set_keepalive_full()` is not ideal.

Of your suggestions, I would think (2) is better, as it does not limit users to the options that we decide need abstraction/mapping.

Providing this option does not require users to use it, or make the library any less user-friendly. Anyone who wants to set these options will probably have a deep enough understanding of networking and how to implement it using (2) anyway - and if they don't, then perhaps they shouldn't be doing so ;). Personally I would find (1) frustrating having to look up the mapping to libstrophe's own names for the options I want to set.

If there are socket options that should be set by default to be more user-friendly, then that should be done. Maybe `xmpp_conn_set_keepalive()` (or other name to avoid breaking API) could be a helper to turn keepalive functionality on or off, with sensible defaults (and set TCP_USER_TIMEOUT=TCP_KEEPIDLE + TCP_KEEPINTVL * TCP_KEEPCNT), for users without intimate knowledge. But I think that's a slightly separate issue.

We might want a bit more context in the callback than just the socket, so passing xmpp_conn too might be helpful.

```c
/** `sock` is of type `SOCKET` on Win and `int` on the others */
typedef int (*xmpp_sockopt_cb)(xmpp_conn_t *conn, void *sock);
void xmpp_conn_set_sockopt_callback(xmpp_conn_t *conn, xmpp_sockopt_cb *cb);
```

Reply:
> We might want a bit more context in the callback than just the socket, so passing xmpp_conn too might be helpful.

true! I forgot to add the `conn` in the example.

```c
/** The function which will be called when Strophe creates a new socket.
 *
 *  The `sock` argument is a pointer that is dependent on the architecture
 *  Strophe is compiled for.
 *
 *  For POSIX compatible systems usage shall be:
 *  ```
 *  int soc = *((int*)sock);
 *  ```
 *
 *  On Windows usage shall be:
 *  ```
 *  SOCKET soc = *((SOCKET*)sock);
 *  ```
 *
 *  This function will be called for each socket that is created, right after
 *  the values set via `xmpp_conn_set_keepalive()` are applied.
 *
 *  OR
 *
 *  This function will be called for each socket that is created and in case
 *  this call-back is set, the values set via `xmpp_conn_set_keepalive()` are
 *  ignored.
 *
 *
 *  `examples/complex.c` implements some of the settings that could be useful
 *  for an application.
 *
 *
 *  @param conn     The Strophe connection object this callback originates from.
 *  @param sock     A pointer to the underlying file descriptor.
 *
 *  @return 0 on success, -1 on error
 *
 *  @ingroup Connections
 */
typedef int (*xmpp_sockopt_callback)(xmpp_conn_t *conn, void *sock);
void xmpp_conn_set_sockopt_callback(xmpp_conn_t *conn, xmpp_sockopt_callback *cb);
```

In the API documentation above I've added two options we have to choose from, either apply the values set via `xmpp_conn_set_keepalive()` before calling the callback or ignoring those values ... I'm slightly in favor of the latter as it would result in clear separation of who's responsible for socket options: either libstrophe or the User and not a mix of both ... but there's also the user-friendliness aspect where one could argue for the former ... :smile: 

> If there are socket options that should be set by default to be more user-friendly, then that should be done. Maybe `xmpp_conn_set_keepalive()` (or other name to avoid breaking API) could be a helper to turn keepalive functionality on or off, with sensible defaults (and set TCP_USER_TIMEOUT=TCP_KEEPIDLE + TCP_KEEPINTVL * TCP_KEEPCNT), for users without intimate knowledge. But I think that's a slightly separate issue.

I like the idea, but I'd prefer to not add more API's if we add a generic solution now. Especially since this seems to be a linux-specific thing.

With the existing API we already have `TCP_USER_TIMEOUT` and `TCP_KEEPINTVL` of your equation ... we could choose some sensible default for `TCP_KEEPIDLE` or `TCP_KEEPCNT` and calculate the other at runtime!? Then simply add those settings as default behavior!? This shouldn't break anything or not?

**but** combined with the choice above, I guess we should go with

* keep `xmpp_conn_set_keepalive()` behavior as is
* call callback after values are applied

This would be the least change for the user from a compatibility PoV, even though functionality wise it's less than what we could achieve ... ah, breaking/changing existing functionality ... what a pleasure to think it through :joy:

Reply:
See #202 🙂